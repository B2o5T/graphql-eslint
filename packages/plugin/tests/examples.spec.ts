import { spawnSync } from 'child_process';
import { relative, join } from 'path';
import type { ESLint } from 'eslint';

const ROOT_CWD = process.cwd()

function countErrors(results: ESLint.LintResult[]): number {
  return results.reduce<number>((acc, curr: { fatalErrorCount: number } & ESLint.LintResult) => {
    if (curr.fatalErrorCount > 0) {
      throw new Error('Found fatal error');
    }
    return acc + curr.errorCount;
  }, 0);
}

function getESLintOutput(cwd: string): ESLint.LintResult[] {
  const { stdout, stderr } = spawnSync('eslint', ['.', '--format', 'json'], { cwd });
  const errorOutput = stderr.toString();
  if (errorOutput) {
    throw new Error(errorOutput);
  }
  const output = stdout.toString();
  const start = output.indexOf('[{');
  const end = output.lastIndexOf('}]') + 2;
  return JSON.parse(output.slice(start, end));
}

function testSnapshot(results: ESLint.LintResult[]): void {
  const normalizedResults = results
    .map(result => ({
      filePath: relative(ROOT_CWD, result.filePath),
      messages: result.messages,
    }))
    .filter(result => result.messages.length > 0);

  expect(normalizedResults).toMatchSnapshot();
}

describe('Examples', () => {
  it('should work on `.graphql` files', () => {
    const cwd = join(ROOT_CWD, 'examples/basic');
    const results = getESLintOutput(cwd);
    expect(countErrors(results)).toBe(6);
    testSnapshot(results);
  });

  it('should work on `.js` files', () => {
    const cwd = join(ROOT_CWD, 'examples/code-file');
    const results = getESLintOutput(cwd);
    expect(countErrors(results)).toBe(4);
    testSnapshot(results);
  });

  it('should work with `graphql-config`', () => {
    const cwd = join(ROOT_CWD, 'examples/graphql-config');
    const results = getESLintOutput(cwd);
    expect(countErrors(results)).toBe(2);
    testSnapshot(results);
  });

  it('should work with `graphql-config` on `.js` files', () => {
    const cwd = join(ROOT_CWD, 'examples/graphql-config-code-file');
    const results = getESLintOutput(cwd);
    expect(countErrors(results)).toBe(4);
    testSnapshot(results);
  });

  it('should work with `eslint-plugin-prettier`', () => {
    const cwd = join(ROOT_CWD, 'examples/prettier');
    const results = getESLintOutput(cwd);
    expect(countErrors(results)).toBe(23);
    testSnapshot(results);
  });
});
