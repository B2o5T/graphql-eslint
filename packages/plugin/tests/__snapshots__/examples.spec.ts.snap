// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Examples should work on \`.graphql\` files 1`] = `
Array [
  Object {
    filePath: examples/basic/fragment.graphql,
    messages: Array [
      Object {
        message: Fragment named "Test" already defined in:
	fragment2.graphql,
        ruleId: @graphql-eslint/unique-fragment-name,
      },
    ],
  },
  Object {
    filePath: examples/basic/fragment2.graphql,
    messages: Array [
      Object {
        message: Fragment named "Test" already defined in:
	fragment.graphql,
        ruleId: @graphql-eslint/unique-fragment-name,
      },
    ],
  },
  Object {
    filePath: examples/basic/query.graphql,
    messages: Array [
      Object {
        message: Anonymous GraphQL operations are forbidden. Please make sure to name your query!,
        ruleId: @graphql-eslint/no-anonymous-operations,
      },
    ],
  },
  Object {
    filePath: examples/basic/schema.graphql,
    messages: Array [
      Object {
        message: Description is required for nodes of type "FieldDefinition",
        ruleId: @graphql-eslint/require-description,
      },
      Object {
        message: Description is required for nodes of type "FieldDefinition",
        ruleId: @graphql-eslint/require-description,
      },
      Object {
        message: Description is required for nodes of type "FieldDefinition",
        ruleId: @graphql-eslint/require-description,
      },
    ],
  },
]
`;

exports[`Examples should work on \`.js\` files 1`] = `
Array [
  Object {
    filePath: examples/code-file/query.js,
    messages: Array [
      Object {
        message: Anonymous GraphQL operations are forbidden. Please make sure to name your query!,
        ruleId: @graphql-eslint/no-anonymous-operations,
      },
      Object {
        message: Unnecessary "query" suffix in your operation name!,
        ruleId: @graphql-eslint/no-operation-name-suffix,
      },
    ],
  },
]
`;

exports[`Examples should work with \`eslint-plugin-prettier\` 1`] = `
Array [
  Object {
    filePath: examples/prettier/invalid.graphql,
    messages: Array [
      Object {
        message: Delete \`·····\`,
        ruleId: prettier/prettier,
      },
      Object {
        message: Delete \`,\`,
        ruleId: prettier/prettier,
      },
      Object {
        message: Delete \`,\`,
        ruleId: prettier/prettier,
      },
      Object {
        message: Delete \`,\`,
        ruleId: prettier/prettier,
      },
      Object {
        message: Delete \`·\`,
        ruleId: prettier/prettier,
      },
      Object {
        message: Delete \`,\`,
        ruleId: prettier/prettier,
      },
      Object {
        message: Delete \`,\`,
        ruleId: prettier/prettier,
      },
    ],
  },
  Object {
    filePath: examples/prettier/invalid.js,
    messages: Array [
      Object {
        message: Replace \`query·User($userId:······\` with \`⏎··query·User($userId:·\`,
        ruleId: prettier/prettier,
      },
      Object {
        message: Insert \`··\`,
        ruleId: prettier/prettier,
      },
      Object {
        message: Replace \`····id,\` with \`······id\`,
        ruleId: prettier/prettier,
      },
      Object {
        message: Replace \`····name,\` with \`······name\`,
        ruleId: prettier/prettier,
      },
      Object {
        message: Replace \`····isViewerFriend,\` with \`······isViewerFriend\`,
        ruleId: prettier/prettier,
      },
      Object {
        message: Replace \`····profilePicture(size:·50)·\` with \`······profilePicture(size:·50)\`,
        ruleId: prettier/prettier,
      },
      Object {
        message: Insert \`··\`,
        ruleId: prettier/prettier,
      },
      Object {
        message: Insert \`··\`,
        ruleId: prettier/prettier,
      },
      Object {
        message: Insert \`··\`,
        ruleId: prettier/prettier,
      },
      Object {
        message: Insert \`··\`,
        ruleId: prettier/prettier,
      },
      Object {
        message: Insert \`··\`,
        ruleId: prettier/prettier,
      },
      Object {
        message: Replace \`··uri,\` with \`····uri\`,
        ruleId: prettier/prettier,
      },
      Object {
        message: Replace \`··width,\` with \`····width\`,
        ruleId: prettier/prettier,
      },
      Object {
        message: Insert \`··\`,
        ruleId: prettier/prettier,
      },
      Object {
        message: Insert \`··\`,
        ruleId: prettier/prettier,
      },
      Object {
        message: Insert \`;\`,
        ruleId: prettier/prettier,
      },
    ],
  },
]
`;

exports[`Examples should work with \`graphql-config\` 1`] = `
Array [
  Object {
    filePath: examples/graphql-config/operations/query.graphql,
    messages: Array [
      Object {
        message: Anonymous GraphQL operations are forbidden. Please make sure to name your query!,
        ruleId: @graphql-eslint/no-anonymous-operations,
      },
    ],
  },
  Object {
    filePath: examples/graphql-config/operations/user.fragment.graphql,
    messages: Array [
      Object {
        message: Field "name" defined multiple times.,
        ruleId: @graphql-eslint/avoid-duplicate-fields,
      },
    ],
  },
]
`;

exports[`Examples should work with \`graphql-config\` on \`.js\` files 1`] = `
Array [
  Object {
    filePath: examples/graphql-config-code-file/query.js,
    messages: Array [
      Object {
        message: Anonymous GraphQL operations are forbidden. Please make sure to name your query!,
        ruleId: @graphql-eslint/no-anonymous-operations,
      },
      Object {
        message: Field "id" must be selected when it's available on a type. Please make sure to include it in your selection set!
If you are using fragments, make sure that all used fragments  specifies the field "id".,
        ruleId: @graphql-eslint/require-id-when-available,
      },
    ],
  },
]
`;
